docs/syntax-named-function-type-returns.md

```diff
@@ -0,0 +1,203 @@
+
+```Lua
+local function doStuff(): (red: number, blue: number)
+    -- ...
+end
+
+local function receiveStuff(blue: number, red: number)
+    -- ...
+end
+
+-- lint: `doStuff()` returns identically-named values in a different order, change these names to silence
+local blue, red = doStuff()
+
+-- lint: `doStuff` sends identically named & typed values to `receiveStuff`, but in the wrong order
+receiveStuff(doStuff())
+```
+
+## Drawbacks
+
+There is a philosophical disagreement over the purpose of the Luau static type system:
+- This proposal was written on the grounds that Luau should have internal syntactic consistency and cross-pollination. Runtime, type checking, and documentation features should enmesh and enrich each other as a unified whole, to ensure everything is kept in sync and is easily comprehensible.
+- The primary argument against this proposal is the belief that Luau's type checker should be a "pure proof system", cleanly separated from documentation concerns at every point, so that the type system can exist in a pure realm unconcerned with semantic information.
+
+This feels like a core tension about the direction of type checking features as a whole. We should check in with the community on this.
+
+There is a particular concern that introducing comprehension aids into the type language would imply meaning where there is none. Would users think that return names have functional meaning?
+
+There are other places in Luau where we have previously introduced comprehension aids in the past:
+
+```Lua
+-- number literals could have been kept pure, but we chose to add meaningless delimiters for easier comprehension
+local ONE_BILLION = 1000000000
+local ONE_BILLION = 1_000_000_000
+
+-- generics could have been referenced abstractly, but we chose to give them human-readable names
+type Foo<2> = (<1>, <2>) -> <1> | <2>
+type Foo<OK, Fallback> = (OK, Fallback) -> OK | Fallback
+
+-- parameters in function types didn't have to have named parameters, but we decided to add them
+type Func = (foo: number, bar: number) -> ()
+```
+
+In particular, it is established convention that Luau does not rearrange the contents of lists bounded by parentheses, even when list items are named:
+
+```Lua
+local function poem(red: number, blue: number)
+    print("roses are", red, "violets are", blue)
+end
+
+local blue, red = "blue", "red"
+
+poem(blue, red) --> roses are blue violets are red
+```
+
+The same principle also applies to multiple assignment in existing parts of Luau:
+
+```Lua
+for value, key in pairs({"hello", "world"}) do
+    print(value) --> 1 2
+end
+```
+
+So, this proposal posits that there is already established precedent for such features, and that users understand how comprehension aids function in Luau today.
+
+A common concern is whether these comprehension aids would mislead people into believing that names are significant when considering type compatibility.
```

Comment:
I'm not sure what mismatch you mean here. I noticed the example wouldn't typecheck because `maybeMap` should prob return `U?` rather than `U`, but that's not got anything to do with return names.

Reply:
> I'm not sure what mismatch you mean here.

The signature of `maybeMap` states it takes a function `f` of type `(item: T) -> (result: U)`, but when we invoke it, we do so with `blah` who's type is `(foo: number) -> (bar: string)`. The mismatch is in the names of the params / return types.

> I noticed the example wouldn't typecheck because maybeMap should prob return U? rather than U, but that's not got anything to do with return names.

Yeah that has to do with me making a small mistake 😅

Reply:
Ah I see. Sometimes I do this intentionally, where e.g. a generic `maybeMap` function uses generic naming, but a specific map function might use names more descriptive based on local information, e.g.:

```
local maybeChar = tryGetCharacter()

local maybeHumanoid = maybeMap(maybeChar, function(char)
   return char:FindFirstChildWhichIsA("Humanoid")
end)
```

So I wouldn't think a discrepancy like that is always a problem.

Reply:
That's my point: unless you _have_ named parameters, people don't try to match the naming.

OCaml has named parameters that start with tildes, they're a part of the type of the function:
```OCaml
(* math.mli *)
val divide : ~dividend:int -> ~divisor:int -> int
(* math.ml *)
let divide ~dividend ~divisor = dividend / divisor
let halve number = divide ~dividend:number ~divisor:2
```
There's support for punning, so one can write code like:
```Ocaml
let f ast_node = (* do something with an ast_node here *) in
(* `~f` is both the named parameter _and_ the local variable *)
List.map ~f list_of_ast_nodes 
```

Reply:
Right. So I think that starts to drift into somewhat adjacent territory around things like named parameter support (as opposed to today's positional parameters) which is out of scope since that's a much heavier change.

I would think of both argument names and return names as "sensible defaults" which document the type generically, without restricting how they're used downstream - it manifests mostly in LSP features.